crates/host/src/module.rs

```diff
@@ -320,3 +396,90 @@ impl PlruCache for InstanceCache {
         &mut self.cache
     }
 }
+
+#[derive(Debug)]
+pub struct ModuleCache {
+    serialized_module_cache: Arc<RwLock<SerializedModuleCache>>,
+    deserialized_module_cache: Arc<RwLock<DeserializedModuleCache>>,
+}
+
+impl ModuleCache {
+    pub fn new(maybe_fs_dir: Option<PathBuf>) -> Self {
+        let serialized_module_cache = Arc::new(RwLock::new(
+            SerializedModuleCache::default_with_engine(make_compiler_engine, maybe_fs_dir),
+        ));
+        let deserialized_module_cache = Arc::new(RwLock::new(DeserializedModuleCache::default()));
+        ModuleCache {
+            serialized_module_cache,
+            deserialized_module_cache,
+        }
+    }
+
+    pub fn get(&self, key: CacheKey, wasm: &[u8]) -> Result<Arc<Module>, wasmer::RuntimeError> {
+        // check deserialized module cache first for module
+        {
+            let mut deserialized_cache = self.deserialized_module_cache.write();
+            if let Some(module) = deserialized_cache.get_item(&key) {
+                return Ok(module);
+            }
+        }
+        // get module from serialized cache otherwise
+        // if cache does not contain module, it will be built from wasm bytes
+        // and then cached in serialized cache
+        let module;
+        {
+            let mut serialized_cache = self.serialized_module_cache.write();
+            module = serialized_cache.get(key, wasm)?;
+        }
+        // cache in deserialized module cache too
+        {
+            let mut deserialized_cache = self.deserialized_module_cache.write();
+            deserialized_cache.put_item(key, module.clone());
+        }
+        Ok(module)
+    }
+}
+
+#[test]
+fn cache_test() {
```

Comment:
Should these tests be in a module that is only available in test config?

Reply:
Also, YAY tests! :)

Reply:
I wonder too

Reply:
Ahh okay, again this is an existing pattern for this create. Understood, something for another time!

Reply:
@ThetaSinner @jost-s uh no? there are tests in the `test` folder, why is this here? 😅

Reply:
@jost-s ah, so then can we put it behind a cfg?

Reply:
@thedavidmeister done